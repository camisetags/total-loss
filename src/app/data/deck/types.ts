export interface Deck {
  id: number;
  selected: boolean;
  name: string;
  description: string;
  cards: object[];
}

export type DeckState = {
  [id: number]: Deck;
};
