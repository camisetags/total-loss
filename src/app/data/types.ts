import { Actions as DeckActions } from './deck/actionTypes';
import { DeckState } from './deck/types';

export type RootAction = DeckActions[keyof DeckActions];

export type RootState = {
  deck: DeckState;
};
