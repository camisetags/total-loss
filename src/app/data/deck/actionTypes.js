export const LIST_DECKS = 'LIST_DECKS';
export const SELECT_DECK = 'SELECT_DECK';
