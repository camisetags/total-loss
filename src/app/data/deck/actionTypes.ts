import { Deck } from './types';

export const LIST_DECKS = 'LIST_DECKS';
export const SELECT_DECK = 'SELECT_DECK';

export type Actions = {
  LIST_DECKS: {
    type: typeof LIST_DECKS;
    data: any;
  };
  SELECT_DECK: {
    type: typeof SELECT_DECK;
    data: any;
  };
};
