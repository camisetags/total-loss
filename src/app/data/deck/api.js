import requestHandler from '../requestHandler';

const DeckApi = {
  getList: () => requestHandler.get('/api/deck/'),
  getById: id => requestHandler.get(`/api/deck/${id}`),
};

export default DeckApi;
