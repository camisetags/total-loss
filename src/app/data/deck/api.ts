import requestHandler from '../requestHandler';
import { Api } from '../types';

const DeckApi: Api = {
  getList: () => requestHandler.get('/api/deck/'),
  getById: (id: number) => requestHandler.get(`/api/deck/${id}`),
};

export default DeckApi;
