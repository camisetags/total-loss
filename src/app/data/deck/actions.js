import * as deckTypes from './actionTypes';
import deckApi from './api';

export const getDeckList = () => async (dispatch) => {
  const response = await deckApi.getList();
  dispatch({
    type: deckTypes.LIST_DECKS,
    data: response.data.results,
  });
};

export const selectDeck = deckId => async (dispatch) => {
  const response = await deckApi.getById(deckId);
  dispatch({
    type: deckTypes.SELECT_DECK,
    data: {
      name: response.data.name,
      description: response.data.description,
      cards: [...response.data.cards],
    },
  });
};

export default getDeckList;
