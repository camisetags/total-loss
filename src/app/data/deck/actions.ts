import { normalize } from 'normalizr';
import { Deck } from './types';
import { deckSchema } from './schemas';
import { LIST_DECKS, SELECT_DECK, Actions as DeckActions } from './actionTypes';
import deckApi from './api';

import { Dispatch } from 'redux';

export const getDeckList = () => async (dispatch: Dispatch<DeckActions>) => {
  const response = await deckApi.getList();
  const deckList = response.data.results.map((deck: Deck) => ({
    ...deck,
    selected: false,
  }));

  dispatch({
    type: LIST_DECKS,
    data: normalize(deckList, [deckSchema]).entities.decks,
  });
};

export const selectDeck = (deckId: number) => async (dispatch: Dispatch<DeckActions>) => {
  const response = await deckApi.getById(deckId);
  const deck = {
    id: deckId,
    selected: true,
    name: response.data.name,
    description: response.data.description,
    cards: [...response.data.cards],
  };

  dispatch({
    type: SELECT_DECK,
    data: normalize(deck, deckSchema).entities.decks,
  });
};

export default getDeckList;
