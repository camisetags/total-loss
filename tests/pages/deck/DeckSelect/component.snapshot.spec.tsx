import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { MemoryRouter, Link } from 'react-router-dom';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import DeckPageComponent from '../../../../src/app/pages/deck/DeckSelect/component';
import { Deck } from '../../../../src/app/data/deck/types';

describe('DeckSelect page', () => {
  const deckList = [
    {
      id: 1,
      name: 'Testing card id 1',
      description: 'Some testing card',
    },
    {
      id: 2,
      name: 'Testing card id 2',
      description: 'Some testing card',
    },
    {
      id: 3,
      name: 'Testing card id 3',
      description: 'Some testing card',
    },
  ];

  beforeEach(() => {
    Enzyme.configure({ adapter: new Adapter() });
  });

  it('renders DeckList page component', () => {
    const mockFunc = jest.fn();
    const rendered = renderer.create(
      <MemoryRouter>
        <DeckPageComponent decks={deckList as Deck[]} selectDeck={() => mockFunc()} />
      </MemoryRouter>,
    );

    expect(rendered).toMatchSnapshot();
  });

  it('should call selectDeck event', () => {
    const mockFunc = jest.fn();

    const pageWrapper = mount(
      <MemoryRouter>
        <DeckPageComponent decks={deckList as Deck[]} selectDeck={deckId => mockFunc(deckId)} />
      </MemoryRouter>,
    );

    pageWrapper
      .find(Link)
      .at(0)
      .simulate('click');

    expect(mockFunc.mock.calls.length).toBe(1);
  });
});
