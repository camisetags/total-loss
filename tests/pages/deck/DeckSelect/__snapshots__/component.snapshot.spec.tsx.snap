// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckSelect page renders DeckList page component 1`] = `
<div>
  <h1
    className="main-title"
  >
    Escolha um dos Decks
  </h1>
  <div
    className="cards-container"
  >
    <div
      className="card card-center"
    >
      <h1>
        Testing card id 1
      </h1>
      <p>
        Some testing card
      </p>
      <hr />
      <a
        className="button small"
        href="players-select"
        onClick={[Function]}
      >
        Escolher
      </a>
    </div>
    <div
      className="card card-center"
    >
      <h1>
        Testing card id 2
      </h1>
      <p>
        Some testing card
      </p>
      <hr />
      <a
        className="button small"
        href="players-select"
        onClick={[Function]}
      >
        Escolher
      </a>
    </div>
    <div
      className="card card-center"
    >
      <h1>
        Testing card id 3
      </h1>
      <p>
        Some testing card
      </p>
      <hr />
      <a
        className="button small"
        href="players-select"
        onClick={[Function]}
      >
        Escolher
      </a>
    </div>
  </div>
</div>
`;
